tests/basic_event_manager.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+// A basic, single threaded application that only needs one type of
+// subscriber: `CounterSubscriber`.
+//
+// The application has an `EventManager` and can register multiple subscribers
+// of type `CounterSubscriber`.
+extern crate event_manager;
+extern crate vmm_sys_util;
+
+mod common;
+
+use common::CounterSubscriber;
+use event_manager::{EventManager, SubscriberId, SubscriberOps};
+use std::ops::Drop;
+
+struct App {
+    event_manager: EventManager<CounterSubscriber>,
+    subscribers_id: Vec<SubscriberId>,
+}
+
+impl App {
+    fn new() -> Self {
+        Self {
+            event_manager: EventManager::<CounterSubscriber>::new().unwrap(),
+            subscribers_id: vec![],
+        }
+    }
+
+    fn add_subscriber(&mut self) {
+        let counter_subscriber = CounterSubscriber::new();
+        let id = self.event_manager.add_subscriber(counter_subscriber);
+        self.subscribers_id.push(id);
+    }
+
+    fn run(&mut self) {
+        let _ = self.event_manager.run_with_timeout(100);
+    }
+
+    fn inject_events_for(&mut self, subscriber_index: &[usize]) {
+        for i in subscriber_index {
+            let subscriber = self
+                .event_manager
+                .subscriber_mut(self.subscribers_id[*i])
+                .unwrap();
+            subscriber.trigger_event();
+        }
+    }
+
+    fn clear_events_for(&mut self, subscriber_index: &Vec<usize>) {
+        for i in subscriber_index {
+            let subscriber = self
+                .event_manager
+                .subscriber_mut(self.subscribers_id[*i])
+                .unwrap();
+            subscriber.clear_event();
+        }
+    }
+
+    fn get_counters(&mut self) -> Vec<u64> {
+        let mut result = Vec::<u64>::new();
+        for subscriber_id in &self.subscribers_id {
+            let subscriber = self.event_manager.subscriber_mut(*subscriber_id).unwrap();
+            result.push(subscriber.counter());
+        }
+
+        result
+    }
+}
+
+impl Drop for App {
+    fn drop(&mut self) {
```

Comment:
I think the subscribers within the event manager get dropped/removed automatically when the `App` object is dropped (because the `event_manager` member will be dropped, which drops the `EventManager::subscribers` collection and its members). Is there any reason for explicitly implementing `Drop` in this example?

Reply:
`EventManager` does not implement `Drop` (maybe it should), but this is not the point. What I am trying to exemplify here is that once you are done with the subscriber, you need to call remove. Remove is not just removing things from the vector collection (which is something that rust does anyway), but also calls `epoll_ctl_del` to let `epoll` know that it doesn't need to trigger events for those fds. Does it make sense?

Reply:
Sure makes sense; maybe some comments in the code can help drive this point further. The example itself is a bit contrived because all these things happen by default (i.e. without an explicit `Drop` implementation for `App`), but explanations add any potentially missing context. On a related note, regardless of whether a type has  a `Drop` implementation  or not, its members are recursively dropped when a value of that type goes out of scope anyway (some more info [here](https://doc.rust-lang.org/std/ops/trait.Drop.html)).

Reply:
Maybe move the `for` loop that removes subscribers in a non-`Drop` function (`App::stop_listening()` or something like that). Make it explicit that `remove_subscribers` is necessary (recommended?) when you don't care about those events anymore - maybe your app has moved on to a stage where events are irrelevant.

Reply:
I moved the for loop to a different function, added more comments to what's the deal with it. This conversation is also a bit confusing to me, so I am not sure that's what you were asking for or not :))

Please take another look.